crates/ruff_server/src/server/api/requests/code_action.rs

```diff
@@ -1,81 +1,222 @@
+            title: format!("{DIAGNOSTIC_NAME} ({code}): {title}"),
+            kind: Some(types::CodeActionKind::QUICKFIX),
+            edit: Some(types::WorkspaceEdit {
+                document_changes: Some(types::DocumentChanges::Edits(edit.document_edits.clone())),
+                ..Default::default()
+            }),
+            ..Default::default()
+        })
+    })
+}
+
+impl SupportedCodeAction {
+    fn kind(self) -> CodeActionKind {
+        match self {
+            Self::QuickFix => CodeActionKind::QUICKFIX,
+            Self::FixAll => CodeActionKind::SOURCE_FIX_ALL,
+            Self::OrganizeImports => CodeActionKind::SOURCE_ORGANIZE_IMPORTS,
+        }
+    }
+}
+
+fn fix_all<'d>(
+    edits: impl Iterator<Item = &'d DiagnosticEdit> + 'd,
+) -> impl Iterator<Item = CodeActionOrCommand> + 'd {
+    let edits_made: Vec<_> = edits
+        .filter(|edit| {
+            edit.diagnostic_fix
+                .fix
+                .applies(ruff_diagnostics::Applicability::Safe)
+        })
+        .collect();
+    let diagnostics_fixed = edits_made
+        .iter()
+        .map(|edit| edit.original_diagnostic.clone())
+        .collect();
+
+    (!edits_made.is_empty())
+        .then(move || {
+            edits_made
+                .into_iter()
+                .flat_map(|edit| edit.document_edits.iter())
+        })
+        .map(|changes| {
+            vec![
+                types::CodeActionOrCommand::CodeAction(types::CodeAction {
+                    title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
+                    diagnostics: Some(diagnostics_fixed),
+                    kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
                     edit: Some(types::WorkspaceEdit {
-                        document_changes: Some(types::DocumentChanges::Edits(changes)),
+                        document_changes: Some(types::DocumentChanges::Edits(
+                            changes.cloned().collect(),
+                        )),
                         ..Default::default()
                     }),
                     ..Default::default()
-                }))
-            })
-            .collect();
+                }), // TODO: implement command handler for the server
+                    /*
+                    types::CodeActionOrCommand::Command(types::Command {
+                        ...
+                    }
+                     */
+            ]
+        })
+        .into_iter()
+        .flatten()
+}
 
-        Ok(Some(
-            actions?
-                .into_iter()
-                .flatten()
-                .map(types::CodeActionOrCommand::CodeAction)
-                .collect(),
-        ))
-    }
+/// If `action_filter` is `None`, this returns the full list of supported code actions. Otherwise,
+/// the list is filtered.
+fn available_code_actions(
+    action_filter: Option<Vec<CodeActionKind>>,
+) -> BTreeSet<SupportedCodeAction> {
```

Comment:
Nit: We could consider using a bitflag for this, considering that the action kinds are a fixed set. But I'm also okay keeping with we have, avoiding the allocation here is probably overkill.